types/tx.go

```diff
@@ -41,3 +45,19 @@ type TxProof struct {
 	Data     Tx               `json:"data"`
 	Proof    merkle.Proof     `json:"proof"`
 }
+
+func (txs Txs) ToTxsWithISRs(intermediateStateRoots IntermediateStateRoots) ([]pb.TxWithISRs, error) {
+	expectedLength := len(txs) + 2
+	if len(intermediateStateRoots.RawRootsList) != len(txs)+2 {
+		return nil, fmt.Errorf("invalid length of ISR list: %d, expected length: %d", len(intermediateStateRoots.RawRootsList), expectedLength)
+	}
+	txsWithISRs := make([]pb.TxWithISRs, 0)
+	for i, tx := range txs {
+		txsWithISRs = append(txsWithISRs, pb.TxWithISRs{
+			PreIsr:  intermediateStateRoots.RawRootsList[i],
```

Comment:
what is the implicit assumption here? first isr in the list corresponds to previous tx isr and last isr in the list corresponds to next tx isr?

Reply:
the first ISR corresponds to ISR after a `BeginBlock` call and before the first `DeliverTx` call. Similarly, last ISR corresponds to ISR after `EndBlock` call. Second to last ISR corresponds to ISR after last `DeliverTx` call and before `EndBlock` call.